src/dotnet/commands/dotnet-install/PackageToProjectFileAdder.cs

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.DotNet.ExecutablePackageObtainer;
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.Cli
+{
+
+    internal class PackageToProjectFileAdder : ICanAddPackageToProjectFile
+    {
+        public void Add(FilePath projectPath, string packageId)
+        {
+            if (packageId == null) throw new ArgumentNullException(nameof(packageId));
+
+            var argsToPassToRestore = new List<string>
+            {
+                projectPath.Value,
+                "package",
+                packageId,
+                "--no-restore"
+            };
+
+            var command = new DotNetCommandFactory(alwaysRunOutOfProc: true)
+                .Create(
+                    "add",
+                    argsToPassToRestore)
+                .CaptureStdOut()
+                .CaptureStdErr();
+
+            var result = command.Execute();
+            if (result.ExitCode != 0)
+            {
+                throw new PackageObtainException("Failed to add package. " +
```

Comment:
Rather than formatting the error message at this throw site, could we pass these values as properties of the exception and format it at the catch site?  The command output alone will be multilined, so this will format oddly as is; properties would give the catch site more flexibility in how it is displayed.

Reply:
I intent to just put this as internal exception. The catch side should have no knowledge of detail of this error, just failed to restore package. The catch side should not have the knowledge of how to format the error.

On another side, Either<L,R> would be a much better pattern to provide rich information. But it is not well adapted.

Reply:
I don't see how this invalidates @peterhuene 's comment. Just push the formatting of the message to the constructor.

However is catching this is still displaying the message, no?

Reply:
There are several topic in this thread I think, however I don't want to get into exception handling style discussion

Format in the exception constructor: I agree, however, after loc, this will becomes a string.format, so I think it will be a throw away work